clickhouse: add cookbook recipe on integrating kafka and clickhouse with terraform (#1586)
aiven/devportal: docs/tools/terraform/reference/cookbook/kafka-clickhouse-integration-recipe.rst

Aiven for Apache Kafka® as a source for Aiven for ClickHouse®
=============================================================

This article shows by way of example how to integrate Aiven for Apache Kafka® with Aiven for ClickHouse® using `Terraform provider for Aiven <https://registry.terraform.io/providers/aiven/aiven/latest/docs>`_. An Apache Kafka® source topic is used as a data source and Aiven for ClickHouse® is used to filter or transform the raw data with a materialized view before writing it to a regular table.

> 

Reviewer: Let's give an example of the sensor data, it will be useful to understand the ``clickhouse_kafka_user_config`` terraform block below. This could be for instance:

```suggestion
.. code-block:: javascript

  {
    "sensor_id": 10000001,
    "ts": "2022-12-01T10:08:24.446369",
    "key": "cpu_usage",
    "value": 96.42
  }
```

I don't know how that renders, or if you prefer to use another language in the ``code-block`` directive.

Reviewer: The suggestion should be a bit lower in the narrative, e.g. around line 10.

       }

  3. ``*.tfvars`` file

  Use it to indicate the actual values of the variables so that they can be passed (with the ``-var-file=`` flag) to Terraform during runtime and excluded later on. Configure the ``var-values.tfvars`` file as follows:

    .. code-block:: terraform

       aiven_api_token     = "<YOUR-AIVEN-AUTHENTICATION-TOKEN-GOES-HERE>"
       project_name        = "<YOUR-AIVEN-CONSOLE-PROJECT-NAME-GOES-HERE>"

Configure the ``services.tf`` file
''''''''''''''''''''''''''''''''''

The following Terraform script initializes both Aiven for Apache Kafka and Aiven for ClickHouse services, creates the service integration, the source Apache Kafka topic, and the Aiven for ClickHouse database.

.. code-block:: terraform

  resource "aiven_kafka" "kafka" {
    project                 = var.project_name
    cloud_name              = "google-europe-west1"
    plan                    = "business-4"
    service_name            = "kafka-gcp-eu"
    maintenance_window_dow  = "monday"
    maintenance_window_time = "10:00:00"
  }

  resource "aiven_kafka_topic" "source" {
    project      = var.project_name
    service_name = aiven_kafka.kafka.service_name

>     partitions   = 2

Reviewer: We can increase the number of partitions to make it more realistic
```suggestion
    partitions   = 50
```

    replication  = 3
    topic_name   = "iot_measurements"
  }

  resource "aiven_clickhouse" "clickhouse" {
    project                 = var.project_name
    cloud_name              = "google-europe-west1"
    plan                    = "startup-beta-8"
    service_name            = "clickhouse-gcp-eu"
    maintenance_window_dow  = "monday"
    maintenance_window_time = "10:00:00"
  }

  resource "aiven_service_integration" "clickhouse_kafka_source" {
    project                  = var.project_name
    integration_type         = "clickhouse_kafka"
    source_service_name      = aiven_kafka.kafka.service_name

>     destination_service_name = aiven_clickhouse.clickhouse.service_name

Reviewer: This is not complete, to enable the integration user need to specify the target table schema like so
```suggestion
    destination_service_name = aiven_clickhouse.clickhouse.service_name
    clickhouse_kafka_user_config {
      tables {
	name        = "edge_measurements_raw"
	group_name  = "clickhouse-ingestion"
	data_format = "JSONEachRow"
	columns {
	  name = "sensor_id"
	  type = "UInt64"
	}
	columns {
	  name = "ts"
	  type = "DateTime64(6)"
	}
	columns {
	  name = "key"
	  type = "LowCardinality(String)"
	}
	columns {
	  name = "value"
	  type = "Float64"
	}
	topics {
	  name = aiven_kafka_topic.edge_measurements.topic_name
	}
      }
    }

```

  1. Run the following command:

    .. code-block:: shell

       terraform init
  
  The ``init`` command performs initialization operations to prepare the working directory for use with Terraform. For this recipe, ``init`` automatically finds, downloads, and installs the necessary Aiven Terraform Provider plugins.

  2. Run the following command:

    .. code-block:: bash

       terraform plan -var-file=var-values.tfvars
  
  The ``plan`` command creates an execution plan and shows the resources to be created (or modified). This command doesn't actually create any resources but gives you a heads-up on what's going to happen next.

  3. If the output of ``terraform plan`` looks as expected, run the following command:

    .. code-block:: bash

       terraform apply -var-file=var-values.tfvars
  
  The ``terraform apply`` command creates (or modifies) your infrastructure resources.

Check out the results
---------------------

* Resource ``aiven_clickhouse`` creates an Aiven for ClickHouse service with the project name, the cloud name (provider, region, zone), the Aiven service plan, and the service name as specified in the ``services.tf`` file.

> * Resource ``aiven_clickhouse_database`` resources creates a database for writing raw Kafka messages, where new tables and views are created based on the processed messages.

Reviewer: The database is separate from the landing DB for the service integration, so I would propose

```suggestion
* Resource ``aiven_clickhouse_database`` resources creates a database that can be used to further transform the ingested data and perform analytics on it
```

* Resource ``aiven_kafka`` creates an Aiven for Apache Kafka cluster.
* Resource ``aiven_kafka_topic`` creates Apache Kafka topic ``iot_measurements``.

> * Resource ``aiven_service_integration`` resource creates the integration between the Aiven for Apache Kafka and the Aiven for ClickHouse service.

Reviewer: The servic integration creates a database to insert the ingested data to. In this instance, the database name will be ``service_kafka-gcp-eu`` (it depends on the kafka service name)
